flink-cdc-common/src/main/java/org/apache/flink/cdc/common/sink/HashFunctionProvider.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.cdc.common.sink;
+
+import org.apache.flink.cdc.common.event.TableId;
+import org.apache.flink.cdc.common.schema.Schema;
+
+import java.io.Serializable;
+
+/**
+ * Provide {@link HashFunction} to help PrePartitionOperator to shuffle DataChangeEvent to
+ * designated subtask. This is usually beneficial for load balancing, when writing to different
+ * partitions/buckets in {@link DataSink}, add custom Implementation to further improve efficiency.
+ */
+public interface HashFunctionProvider extends Serializable {
+    HashFunction getHashFunction(TableId tableId, Schema schema);
+
+    // --------------------------------------------------------------------------------------------
+    //  Default life cycle methods
+    // --------------------------------------------------------------------------------------------
+    default void open() {}
+
+    default void close() {}
```

Comment:
Since we provided the `getHashFunction` method based on `TableId`, connector implementers might use `TableId` to obtain the actual schema of the database and perform some caching operations. We can establish connections and initialize caches in the open method. These two lifecycle methods were added following @lvyanquan 's suggestion, and he might respond with further additions.

Reply:
Usually, it will need to obtain partition or bucket information based on TableId.
I wonder if it is possible to cache catalog or connection here to reuse objects or connections.

Reply:
A function with lifecycle management makes sense to me, why the factory need to care the resource management? could we push the logic that `use TableId to obtain partition or buckets info` to `HashFunction` internal? The HashFunction need to open/close  may be better, what do you think?

Reply:
Actually, HashFunctionalProvider is similar to the role of a `catalog`, and HashFunction is similar to the role of a `table`, so the open/lose method is called on the `catalog` role.

Reply:
Well, Catalog is used as a metadata manager instance instead of a `table` factory,  `table` is not constructed by the instance. A metadata manager has its life cycle makes sense to me, a factory own lifecycle confuse me a little. Come back to the function itself, Flink also has many functions as well as the function factories, I didn't see the necessary why a function factory need to care the required resource in runtime, but function manages its resources is pretty common.

Reply:
You can also visit flink code about:

- RichFunction & SourceFunctionProvider

- InputFormat & InputFormatProvider

Reply:
So the key point is that what we need is a metadata manager or a function factory. I agree that what we need is one function factory, though it may need the assistance of specific metadata manager.

And the resource utilization is indeed overthinking, because we can release database connection after extracting the parameters for calculation.

Reply:
Yeah, that's what I want to propose

Reply:
In my opinion, I believe that HashFunction should not have lifecycle functions. It is merely a hash function, not a Flink operator, and it will be cached and disappear automatically. Its lifecycle can be perceived as a kind of "constant" after creation, so we should not concern ourselves with its lifecycle.

Based on this viewpoint, whether HashFunctionProvider should have lifecycle functions depends solely on whether we need to reuse some resources created at runtime when creating a HashFunction.

In fact, I also think that currently, the connectors do not have any tasks that need to be done within lifecycle functions. Therefore, I have removed the lifecycle functions for now.